### PR TITLE
Compatibility with compiler v11 and uncurried mode.

### PR DESCRIPTION
**Pull request by cristianoc**

### CODE REVIEW

**Comment by cristianoc** [comment]
Reformat would undo some changes, as it blurs the arity.
One could do this to prevent the blurring: https://github.com/cristianoc/rescript-json-combinators/pull/1

**Comment by cristianoc** [comment]
Btw I find that currying makes the code much harder to read.
E.g.: this line:
```res
let tuple2 = (decodeA, decodeB, . json) => {
```
really does not convey the intuition: take 2 decoders and return one decoder.

**Comment by glennsl** [comment]
I don't think it's the currying that makes it hard to read, but the formatting enforced by reformat. The formatting you propose here is what I would have wanted. When do you think this would be supported by reformat? Once v11 ships?

**Comment by cristianoc** [comment]
> I don't think it's the currying that makes it hard to read, but the formatting enforced by reformat. The formatting you propose here is what I would have wanted. When do you think this would be supported by reformat? Once v11 ships?

It's supported by v11 in uncurried mode. This is how it's going to look:
```res
let tuple2 = (decodeA, decodeB) => json => {
```
So the dots are gone, and the way one puts things in parens matters.

File: examples/EncodeExample.res [hunk 1]
```diff
@@ -42,8 +42,8 @@ module Encode = {
 
   let line = (~points, ~thickness=?, ()) =>
     Unsafe.object({
-      "points": array(point, points),
-      "thickness": option(int, thickness),
+      "points": array(point)(points),
+      "thickness": option(int)(thickness),
```
Comment (glennsl) [inline examples/EncodeExample.res]: I'm hesitant to change the examples, since this gives the wrong impression of what the API is meant to be. I wonder if it would be better to just make another major release in order to make these uncurried right away instead.

File: src/Json_Decode.res [hunk 2]
```diff
@@ -68,16 +68,16 @@ let array = (decode, . json) => {
       let value = decode(. %raw("json[i]"))
       target->Array.unsafe_set(i, value)
     } catch {
-    | DecodeError(msg) => raise(DecodeError(j`${msg}\n\tin array at index $i`))
+    | DecodeError(msg) => raise(DecodeError(`${msg}\n\tin array at index $i`))
```
Comment (glennsl) [inline src/Json_Decode.res]: Is `$i` now an accepted form of normal interpolated strings, or is this an error that will just print `$i` instead of the value of `i` stringified?

File: src/Json_Encode.res [hunk 3]
```diff
@@ -14,9 +14,9 @@ external boolArray: array<bool> => Js.Json.t = "%identity"
 
 @val external null: Js.Json.t = "null"
 
-let array = (encode, arr) => arr->Js.Array2.map(encode)->jsonArray
+let array = encode => arr => arr->Js.Array2.map(x => encode(x))->jsonArray
```
Comment (glennsl) [inline src/Json_Encode.res:17]: What's the reason for this change? Curried functions can no longer be passed around as first-class values?

**Comment by glennsl** [comment]
> It's supported by v11 in uncurried mode.

Ok, great! I'm sure can keep away from reformat until then.

File: examples/EncodeExample.res (hunk 1 above)
Comment (cristianoc) [reply examples/EncodeExample.res]: If it's not built on npm install, there's no hurry whatsoever for this.

File: src/Json_Decode.res (hunk 2 above)
Comment (cristianoc) [reply src/Json_Decode.res]: `$i` is ignored, as this follows the JS string interpolation spec

Comment (cristianoc) [reply src/Json_Decode.res]: So this is an error

Comment (cristianoc) [reply src/Json_Decode.res]: That's why we recommend to check in the build artifacts

Comment (cristianoc) [reply src/Json_Decode.res]: Fixed now.

File: src/Json_Encode.res (hunk 3 above)
Comment (cristianoc) [reply src/Json_Encode.res:17]: It's a dependency issue. `Js.Array2.map` is curried, so type inference gives a curried type. A type annotation would also work here.

File: examples/EncodeExample.res (hunk 1 above)
Comment (cristianoc) [reply examples/EncodeExample.res]: Removed this change.

A side comment here. One thing I also tested, is what happens when one defines:
```res
@unboxed
type t<'a> = {encode: 'a => Js.Json.t}

```
And this would look like:
```res
array(point).encode(points)
```

Comment (glennsl) [reply examples/EncodeExample.res]: Is there a practical benefit to this? The reason I haven't just made it abstract, like in `Decode`, is that it wouldn't be possible to expose the primitives as `%identity` externals then, meaning it would incur significantly more overhead. And this would have the same effect.

File: src/Json_Encode.res (hunk 3 above)
Comment (glennsl) [reply src/Json_Encode.res:17]: Hmm, I still don't understand what kind of dependency issue would affect inference here. no big deal though, I'm sure I'll find out once v11 drops!

File: examples/EncodeExample.res [hunk 4]
```diff
@@ -6,7 +6,7 @@ let _ = ["foo", "bar"]->Json.Encode.stringArray->Js.log
 
 // Encode string array, slower but more flexible
 // prints ["foo", "bar"]
-let _ = ["foo", "bar"]->Json.Encode.array(Json.Encode.string, _)->Js.log
+let _ = ["foo", "bar"]->Json.Encode.array(Json.Encode.string)(_)->Js.log
```
Comment (glennsl) [inline examples/EncodeExample.res]:
```suggestion
let _ = ["foo", "bar"]->Json.Encode.array(Json.Encode.string, _)->Js.log
```

**Review by glennsl — APPROVED** [review]
